core/core_test.go

```diff
@@ -28,3 +49,151 @@ func TestEthereumChainDetection(t *testing.T) {
+		EthereumRPCMaxContentLength:      524288,
+		EnableEthereumRPCRateLimiting:    false,
+		EthereumRPCMaxRequestsPer24HrUTC: 99999999999999,
+		EthereumRPCMaxRequestsPerSecond:  99999999999999,
+		MaxOrdersInStorage:               100000,
+		CustomOrderFilter:                "{}",
+	}
+	app, err := New(config)
+	require.NoError(t, err)
+	return app
+}
+
+var (
+	rpcClient           *ethrpc.Client
+	ethClient           *ethclient.Client
+	blockchainLifecycle *ethereum.BlockchainLifecycle
+)
+
+// Since these tests must be run sequentially, we don't want them to run as part of
+// the normal testing process. They will only be run if the "--serial" flag is used.
+var serialTestsEnabled bool
+
+func init() {
+	flag.BoolVar(&serialTestsEnabled, "serial", false, "enable serial tests")
+	testing.Init()
+	flag.Parse()
+
+	var err error
+	rpcClient, err = ethrpc.Dial(constants.GanacheEndpoint)
+	if err != nil {
+		panic(err)
+	}
+	ethClient = ethclient.NewClient(rpcClient)
+	blockchainLifecycle, err = ethereum.NewBlockchainLifecycle(rpcClient)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func TestOrderSync(t *testing.T) {
+	if !serialTestsEnabled {
+		t.Skip("Serial tests (tests which cannot run in parallel) are disabled. You can enable them with the --serial flag")
+	}
+
+	teardownSubTest := setupSubTest(t)
+	defer teardownSubTest(t)
+
+	// Set up two Mesh nodes. originalNode starts with some orders. newNode enters
+	// the network without any orders.
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	wg := &sync.WaitGroup{}
+	originalNode := newTestApp(t)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		require.NoError(t, originalNode.Start(ctx))
+	}()
+	newNode := newTestApp(t)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		require.NoError(t, newNode.Start(ctx))
+	}()
+
+	// Manually add some orders to originalNode.
+	originalOrders := make([]*zeroex.SignedOrder, 10)
+	for i := range originalOrders {
+		originalOrders[i] = scenario.CreateWETHForZRXSignedTestOrder(t, ethClient, constants.GanacheAccount1, constants.GanacheAccount2, big.NewInt(20), big.NewInt(5))
+	}
+
+	// We have to wait for latest block to be processed by the Mesh node.
+	time.Sleep(blockProcessingWaitTime)
+
+	results, err := originalNode.orderWatcher.ValidateAndStoreValidOrders(ctx, originalOrders, true, constants.TestChainID)
+	require.NoError(t, err)
+	require.NotEmpty(t, results.Accepted, "tried to add orders but some were invalid: \n%s\n", spew.Sdump(results))
+
+	orderEventsChan := make(chan []*zeroex.OrderEvent)
+	orderEventsSub := newNode.SubscribeToOrderEvents(orderEventsChan)
+	defer orderEventsSub.Unsubscribe()
+
+	// Connect the two nodes *after* adding orders to one of them. This should
+	// trigger the ordersync protocol.
+	err = originalNode.AddPeer(peer.AddrInfo{
+		ID:    newNode.node.ID(),
+		Addrs: newNode.node.Multiaddrs(),
+	})
+	require.NoError(t, err)
+
+	// Wait for newNode to get the orders via ordersync.
```

Comment:
Since we added the order directly via `ValidateAndStoreValidOrders`, we are bypassing GossipSub. I think it is not possible for the order to be shared via GossipSub in this test, but let me know if I'm missing something.

Reply:
Ah yep, that makes sense!